src/Haskell/Ide/Engine/Plugin/ApplyRefact.hs

```diff
@@ -307,7 +307,10 @@ codeActionProvider plId docId _ context = IdeResultOk <$> hlintActions
   where
 
     hlintActions :: IdeM [LSP.CodeAction]
-    hlintActions = catMaybes <$> mapM mkHlintAction (filter validCommand diags)
+    hlintActions = do
+      actions <- mapM mkHlintAction (filter validCommand diags)
+      applyAll <- mkApplyAllAction
```

Comment:
We should only return this hint if there is a hint available to apply
```suggestion
      applyAll <- if null actions then return Nothing else mkApplyAllAction
```

Reply:
The way code actions works is that the server sends diagnostics for the file to the client.  The client then sends back any diagnostics that apply to the source where the cursor is currently located in the file.

So if your cursor is near the end of the file, and there is a hint near the beginning, then the actions will be empty.

Hence my remark in the original description that we should be able to quickly check if there are in fact diagnostics for a given file/diagnostics source combination, anywhere in the file.